### PR TITLE
Add an option to buildx rm to prevent the deletion of the buildkitd d…

### PR DESCRIPTION
**Pull request by zaymat**
…aemon

Add --keep-buildkitd to the `rm` command option to preserve the buildkitd daemon after the buildx context is deleted.

Signed-off-by: Mayeul Blanzat <mayeul.blanzat@datadoghq.com>

### CODE REVIEW

File: driver/docker-container/driver.go [hunk 1]
```diff
@@ -305,20 +305,21 @@ func (d *Driver) Rm(ctx context.Context, force bool, rmVolume bool) error {
...
-					return d.DockerAPI.VolumeRemove(ctx, d.Name+volumeStateSuffix, false)
+		if rmBuildkit {
+			if err := d.DockerAPI.ContainerRemove(ctx, d.Name, dockertypes.ContainerRemoveOptions{
+				RemoveVolumes: true,
+				Force:         force,
+			}); err != nil {
+				return err
+			}
+			for _, v := range container.Mounts {
+				if v.Name == d.Name+volumeStateSuffix {
```
Comment (JulienBalestra) [inline driver/docker-container/driver.go]:
```suggestion
				if v.Name != d.Name+volumeStateSuffix {
					continue
				}
				
```

Comment (JulienBalestra) [reply driver/docker-container/driver.go]: Ideally keep the indent flow reduced.

File: driver/docker/driver.go [hunk 2]
```diff
@@ -33,7 +33,7 @@ func (d *Driver) Stop(ctx context.Context, force bool) error {
 	return nil
 }
 
-func (d *Driver) Rm(ctx context.Context, force bool, rmVolume bool) error {
+func (d *Driver) Rm(ctx context.Context, force bool, rmVolume bool, rmBuildkit bool) error {
```
Comment (JulienBalestra) [inline driver/docker/driver.go]:
```suggestion
func (d *Driver) Rm(ctx context.Context, force, rmVolume, rmBuildkit bool) error {
```

File: driver/kubernetes/driver.go [hunk 3]
```diff
@@ -165,16 +165,18 @@ func (d *Driver) Stop(ctx context.Context, force bool) error {
...
-func (d *Driver) Rm(ctx context.Context, force bool, rmVolume bool) error {
-	if err := d.deploymentClient.Delete(ctx, d.deployment.Name, metav1.DeleteOptions{}); err != nil {
-		if !apierrors.IsNotFound(err) {
-			return errors.Wrapf(err, "error while calling deploymentClient.Delete for %q", d.deployment.Name)
-		}
-	}
-	for _, cfg := range d.configMaps {
-		if err := d.configMapClient.Delete(ctx, cfg.Name, metav1.DeleteOptions{}); err != nil {
+func (d *Driver) Rm(ctx context.Context, force bool, rmVolume bool, rmBuildkit bool) error {
+	if rmBuildkit {
```
Comment (JulienBalestra) [inline driver/kubernetes/driver.go]:
```suggestion
	if !rmBuildkit {
		return nil
	}
```

Comment (JulienBalestra) [reply driver/kubernetes/driver.go]: If I understand correctly, this is doing nothing in this implementation when `rmBuildkit` is false. It's weird to call this method in that case 🤔

File: driver/docker-container/driver.go [hunk 4]
```diff
@@ -305,20 +305,21 @@ func (d *Driver) Rm(ctx context.Context, force bool, rmVolume bool) error {
...
+					if rmVolume {
+						return d.DockerAPI.VolumeRemove(ctx, d.Name+volumeStateSuffix, false)
```
Comment (JulienBalestra) [inline driver/docker-container/driver.go]: Are we sure we return here ?

File: driver/kubernetes/driver.go (hunk 3 above)
Comment (zaymat) [reply driver/kubernetes/driver.go]: legit. I'll make the change

File: driver/docker-container/driver.go (hunk 4 above)
Comment (zaymat) [reply driver/docker-container/driver.go]: FYI this appears as part of my change but it's just a matter of indentation. If that's a real concern, we may need to open a new issue upsteam and create a new PR. 

Though I've taken a look at this function call. The context is there https://github.com/DataDog/buildx/blob/bcfd4348290ecb67122a882e0ef42aa537a51ae6/vendor/github.com/moby/buildkit/util/appcontext/appcontext.go#L17 and does not seems to have timeout. But there may have some hidden timeouts somewhere in the call stack.

File: driver/docker-container/driver.go (hunk 1 above)
Comment (zaymat) [reply driver/docker-container/driver.go]: 👍

File: driver/docker/driver.go (hunk 2 above)
Comment (zaymat) [reply driver/docker/driver.go]: 👍

File: commands/rm.go [hunk 5]
```diff
@@ -70,21 +75,25 @@ func rmCmd(dockerCli command.Cli, rootOpts *rootOptions) *cobra.Command {
...
 }
 
-func rm(ctx context.Context, dockerCli command.Cli, ng *store.NodeGroup, keepState bool) error {
+func rm(ctx context.Context, dockerCli command.Cli, ng *store.NodeGroup, keepState, keepBuildkitd bool) error {
 	dis, err := driversForNodeGroup(ctx, dockerCli, ng, "")
 	if err != nil {
 		return err
 	}
 	for _, di := range dis {
 		if di.Driver != nil {
```
Comment (JulienBalestra) [inline commands/rm.go]:
```suggestion
		if di.Driver == nil {
			continue
		}
```
And de-indent the rest.

**Review by JulienBalestra — APPROVED** [review]

File: commands/rm.go (hunk 5 above)
Comment (zaymat) [reply commands/rm.go]: 👍

**Pull request merged**